### PR TITLE
Add negative filter tests

### PR DESCRIPTION
**Pull request by u9g**

### CODE REVIEW

File: trustfall_core/src/interpreter/execution.rs [hunk 1]
```diff
@@ -241,6 +241,24 @@ fn construct_outputs<'query, AdapterT: Adapter<'query>>(
...
 }
 
+/// takes a FieldValue and produces a usize clamped to 0 and usize::MAX
+fn usize_from_field_value(field_value: &FieldValue) -> Option<usize> {
+    match field_value {
+        FieldValue::Int64(num) => {
+            Some(usize::try_from(*num.max(&0)).expect("i64 can be converted to usize"))
+        }
+        FieldValue::Uint64(num) => {
+            Some(usize::try_from(*num).expect("i64 can be converted to usize"))
```
Comment (u9g) [inline trustfall_core/src/interpreter/execution.rs:251]: usize should be bigger than both of these right? there's no need to also .min this with `usize::MAX` I think

Comment (obi1kenobi) [reply trustfall_core/src/interpreter/execution.rs:251]: It's not actually guaranteed to, it's possible that `usize` is 32-bit e.g. on `wasm32` platforms whereas `u64` is always 64-bit. Your call on whether we sort this out now or not.

Comment (obi1kenobi) [reply trustfall_core/src/interpreter/execution.rs:251]: But either update the docstring or the code, because right now the docstring says this is clamped to `usize::MAX` when the code doesn't actually clamp it.

File: trustfall_core/src/interpreter/execution.rs [hunk 2]
```diff
@@ -241,6 +241,24 @@ fn construct_outputs<'query, AdapterT: Adapter<'query>>(
     }))
 }
 
+/// takes a FieldValue and produces a usize clamped to 0 and usize::MAX
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/execution.rs:244]:
```suggestion
/// Takes a FieldValue and produces a usize clamped between 0 and `usize::MAX`.
```
Docstrings should always be capitalized and use code blocks for code.

File: trustfall_core/src/interpreter/execution.rs [hunk 3]
```diff
@@ -241,6 +241,24 @@ fn construct_outputs<'query, AdapterT: Adapter<'query>>(
...
+        }
+        FieldValue::Uint64(num) => {
+            Some(usize::try_from(*num).expect("i64 can be converted to usize"))
+        }
+        FieldValue::Float64(_)
+        | FieldValue::List(_)
+        | FieldValue::Enum(_)
+        | FieldValue::Boolean(_)
+        | FieldValue::String(_)
+        | FieldValue::Null => None,
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/execution.rs:258]: For `null` it's fine to return `None` (imagine a tag being used), but for everything else we should panic instead because that indicates a bug in the type-checker.

File: trustfall_core/src/interpreter/execution.rs [hunk 4]
```diff
@@ -298,7 +326,7 @@ fn collect_fold_elements<'query, Vertex: Clone + Debug + 'query>(
         // If we are able to pull more than `max_fold_count_limit + 1` elements,
         // we know that this fold is going to get filtered out, so we might as well
         // stop materializing its elements early.
-        let mut fold_elements = Vec::with_capacity(*max_fold_count_limit);
+        let mut fold_elements = Vec::with_capacity((*max_fold_count_limit).min(16));
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/execution.rs:329]: please update the comment to explain why `.min(16)`, it isn't obvious.

File: trustfall_core/test_data/tests/valid_queries/fold_count_filter_eq.graphql.ron
Comment (obi1kenobi) [inline trustfall_core/test_data/tests/valid_queries/fold_count_filter_eq.graphql.ron:1]: Please name the file something that more accurately communicates what's being tested, right now it isn't obvious that the point of the test is to use a negative number as the argument.

File: trustfall_core/test_data/tests/valid_queries/fold_count_filter_eq.graphql.ron [hunk 5]
```diff
@@ -0,0 +1,19 @@
+TestGraphQLQuery (
+    // We use a negative number as a filter to ensure that we properly handle the negative number case in get_max_fold_count_limit.
```
Comment (obi1kenobi) [inline trustfall_core/test_data/tests/valid_queries/fold_count_filter_eq.graphql.ron:2]: A few issues here:
- This line is too long. We don't currently have a linter set up, but please try to keep lines below 100 chars in all files to make sure they are readable.
- The comment references implementation details that may change and that ultimately aren't important. The test ensures that comparisons with negative numbers on count work _in general_, and would catch any failures even if they aren't in `get_max_fold_count_limit()`. I wouldn't name that function or any other implementation details here.
- `get_max_fold_count_limit()` is code and a function. Please try to always put code in backticks and have function names followed by `()` to show they are functions.

Comment (u9g) [reply trustfall_core/test_data/tests/valid_queries/fold_count_filter_eq.graphql.ron]: I spruced up the comments in the tests to look like this:

`// Ensure that we properly handle negative numbers in a `not_one_of` filter on a folded edge's count.`

I believe this is the longest one, and it's 101 characters including `// ` so I think that's fine for now. I removed any mentions to internals in tests, good call.

File: trustfall_core/src/interpreter/execution.rs (hunk 1 above)
Comment (u9g) [reply trustfall_core/src/interpreter/execution.rs:252]: I update3d the docstring for now to just not say it's clamped to usize::MAX, instead it just says that we return a positive usize.

File: trustfall_core/src/interpreter/execution.rs [hunk 6]
```diff
@@ -241,6 +241,27 @@ fn construct_outputs<'query, AdapterT: Adapter<'query>>(
     }))
 }
 
+/// Takes a FieldValue and returns a positive usize. If the FieldValue is a Null,
+/// returns None.
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/execution.rs:245]:
```suggestion
/// Extracts numeric `FieldValue` into a `usize`, clamping negative numbers to 0.
/// Returns `None` on `FieldValue::Null`, and panics otherwise.
```

**Review by obi1kenobi — APPROVED** [review]
